File: src/main/java/kr/devport/api/domain/common/logging/RequestCorrelationFilter.java
package kr.devport.api.domain.common.logging;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.slf4j.MDC;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;

@Component
public class RequestCorrelationFilter extends OncePerRequestFilter {

    @Override
    protected void doFilterInternal(
            HttpServletRequest request,
            HttpServletResponse response,
            FilterChain filterChain
    ) throws ServletException, IOException {
        String requestId = LoggingContext.resolveRequestId(request.getHeader(LoggingContext.REQUEST_ID_HEADER));
        MDC.put(LoggingContext.REQUEST_ID_KEY, requestId);
        response.setHeader(LoggingContext.REQUEST_ID_HEADER, requestId);

        try {
            filterChain.doFilter(request, response);
        } finally {
            MDC.remove(LoggingContext.REQUEST_ID_KEY);
        }
    }
}

File: application-api/src/main/kotlin/kr/devport/api/domain/common/logging/RequestCorrelationFilter.kt
package kr.devport.api.domain.common.logging

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.MDC
import org.springframework.core.Ordered
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
@Order(Ordered.HIGHEST_PRECEDENCE)
class RequestCorrelationFilter : OncePerRequestFilter() {
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        val requestId = LoggingContext.resolveRequestId(request.getHeader(LoggingContext.REQUEST_ID_HEADER))
        MDC.put(LoggingContext.REQUEST_ID_KEY, requestId)
        response.setHeader(LoggingContext.REQUEST_ID_HEADER, requestId)

        try {
            filterChain.doFilter(request, response)
        } finally {
            MDC.remove(LoggingContext.REQUEST_ID_KEY)
        }
    }
}
